src/dsk/alloc.c

```diff
@@ -0,0 +1,34 @@
+#include "alloc.h"
+#include "mdisk.h"
+#include "../mkfs/metaBlocks.h"
+
+static pthread_mutex_t iNodeListMutex = PTHREAD_MUTEX_INITIALIZER;
+const int SLEEP_TIME_IN_SECONDS = 3;
+
+size_t alloc(superBlock* superBlock) {
+    while (superBlock -> free_blklst_lock)
+        sleep(SLEEP_TIME_IN_SECONDS);
+    pthread_mutex_lock(&iNodeListMutex);
+    if (superBlock -> free_blklst_lock) {
+        pthread_mutex_unlock(&iNodeListMutex);
+        return alloc(superBlock);
+    }
+
+    superBlock -> free_blklst_lock = true;
+    superBlock -> num_free_blocks --;
+    size_t new_block = *(superBlock -> free_block_list);
+    superBlock -> free_block_list ++;
+
+    if (superBlock -> num_free_blocks == 0) {
+        superBlock -> free_block_list -= BLOCK_SIZE // is it correct?
+        superBlock -> num_free_blocks = BLOCK_SIZE;
+        disk_block *new_list = (disk_block *)malloc(BLOCK_SIZE);
+        fetchMemoryDiskBlock(new_block, new_list);
```

Comment:
Invoke the method in `dsk/blkfetch.c`, which will decide whether it needs to call the data in memory, or the actual disk.